File: apps/api/pi_dash/core/permissions.py
# Copyright (c) 2023-present Pi Dash Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""Workspace/project membership and role helpers — single source of truth.

These were originally split between ``app/permissions/base.py`` (the
``@allow_permission`` decorator) and ``runner/services/permissions.py``
(workspace helpers). They are consolidated here so the assistant tool layer
can enforce *exactly* the same access control the views enforce, without
coupling ``pi_dash.assistant`` to ``pi_dash.runner``.

Role values come from ``pi_dash.db.models.workspace.ROLE_CHOICES``:
``Admin=20``, ``Member=15``, ``Guest=5``.
"""

from __future__ import annotations

from typing import Iterable, Optional

from pi_dash.db.models import ProjectMember, WorkspaceMember

ROLE_ADMIN = 20
ROLE_MEMBER = 15
ROLE_GUEST = 5


def is_workspace_member(user, workspace_id) -> bool:
    """True if ``user`` is a member (any role) of the given workspace."""
    if user is None or not getattr(user, "is_authenticated", False):
        return False
    return WorkspaceMember.objects.filter(
        workspace_id=workspace_id,
        workspace__platform_access_disabled_at__isnull=True,
        member=user,
        is_active=True,
    ).exists()


def workspace_role(user, workspace_id) -> Optional[int]:
    """Return the user's active role in the workspace, or ``None``."""
    if user is None or not getattr(user, "is_authenticated", False):
        return None
    return (
        WorkspaceMember.objects.filter(
            workspace_id=workspace_id,
            workspace__platform_access_disabled_at__isnull=True,
            member=user,
            is_active=True,
        )
        .values_list("role", flat=True)
        .first()
    )


def workspace_role_by_slug(user, workspace_slug) -> Optional[int]:
    """Return the user's active role in the workspace identified by slug."""
    if user is None or not getattr(user, "is_authenticated", False):
        return None
    return (
        WorkspaceMember.objects.filter(
            workspace__slug=workspace_slug,
            workspace__platform_access_disabled_at__isnull=True,
            member=user,
            is_active=True,
        )
        .values_list("role", flat=True)
        .first()
    )


def is_workspace_admin(user, workspace_id) -> bool:
    """True if ``user`` is an Admin (role >= 20) of the workspace."""
    role = workspace_role(user, workspace_id)
    return role is not None and role >= ROLE_ADMIN


def is_at_least_member(user, workspace_id) -> bool:
    """True if ``user`` is at least Member role (>=15) — not Guest."""
    role = workspace_role(user, workspace_id)
    return role is not None and role >= ROLE_MEMBER


def check_project_role(
    user,
    workspace_slug: str,
    project_id,
    allowed_roles: Iterable[int],
    *,
    allow_workspace_admin_bypass: bool = True,
) -> bool:
    """Mirror of the ``@allow_permission`` PROJECT branch (``app/permissions/base.py``).

    True when the user has an active ``ProjectMember`` row with a role in
    ``allowed_roles`` for the project, OR (when ``allow_workspace_admin_bypass``)
    the user is a project member of any role AND a workspace Admin. This is the
    exact rule the issue/comment write endpoints enforce, so the assistant's
    write tools cannot exceed what the user could do by hand.
    """
    if user is None or not getattr(user, "is_authenticated", False):
        return False

    allowed_values = [int(r) for r in allowed_roles]

    has_allowed_role = ProjectMember.objects.filter(
        member=user,
        workspace__slug=workspace_slug,
        workspace__platform_access_disabled_at__isnull=True,
        project_id=project_id,
        role__in=allowed_values,
        is_active=True,
    ).exists()
    if has_allowed_role:
        return True

    if allow_workspace_admin_bypass:
        is_project_member = ProjectMember.objects.filter(
            member=user,
            workspace__slug=workspace_slug,
            workspace__platform_access_disabled_at__isnull=True,
            project_id=project_id,
            is_active=True,
        ).exists()
        if is_project_member and WorkspaceMember.objects.filter(
            member=user,
            workspace__slug=workspace_slug,
            workspace__platform_access_disabled_at__isnull=True,
            role=ROLE_ADMIN,
            is_active=True,
        ).exists():
            return True

    return False

File: rust-api/crates/auth/src/permissions.rs
#![forbid(unsafe_code)]

//! Permission kernel (F-06): every access rule the Django backend enforces.
//!
//! Python references (one tree replaces all of these):
//!
//! - `pi_dash/app/permissions/` (`base.py`, `workspace.py`, `project.py`,
//!   `page.py`) and its byte-identical copy `pi_dash/utils/permissions/`.
//! - `pi_dash/core/permissions.py` — workspace/project role helpers, the
//!   single source of truth the assistant tool layer shares with the views.
//! - `pi_dash/runner/services/permissions.py` — runner visibility and
//!   management rules (workspace helpers re-exported from core).
//! - `pi_dash/managed_runner/permissions.py` — `IsDesktopSession`.
//! - `pi_dash/license/api/permissions/instance.py` —
//!   `InstanceAdminPermission`.
//!
//! The modules are pure, like the rest of this crate: every
//! `Model.objects.filter(...).exists()` check becomes an explicit boolean
//! (or role) in a `*Facts` struct, and row fetching stays the caller's SQL.
//! Each `decide_*` function additionally takes the request's [`TenantScope`]
//! and denies when the facts were fetched for a different workspace, so a
//! handler cannot authorize outside its workspace even with stale facts —
//! the workspace-scoped handle by construction. (`TenantScope` itself is
//! deny-by-default: only the exact same workspace authorizes.)
//!
//! Ported bugs and divergences (kept byte-for-byte, listed for follow-up):
//!
//! - `app/permissions/base.py` gates the creator bypass on workspace
//!   membership first, while `utils/permissions/base.py` honors it
//!   unconditionally ([`allow::CreatorGate`]).
//! - `WorkspaceOwnerPermission` omits `is_active=True` (inactive admins
//!   pass); every sibling filters active rows.
//! - `WorkspaceViewerPermission` and `WorkspaceUserPermission` are
//!   identical (any active membership).
//! - `WorkSpaceBasePermission` falls off the end (returns `None`, denied)
//!   for unmatched methods; the kernel returns `false` there.
//! - `ProjectPagePermission` uses `Page.objects.get`, so a missing page is
//!   a 404, not a 403 ([`page::PageOutcome::PageNotFound`]).
//! - `can_mutate_states`' workspace-admin override still requires project
//!   membership (a `None` membership returns `false` first).
//! - `check_project_role`'s workspace-admin bypass needs workspace role
//!   exactly `ADMIN` (`role=ROLE.ADMIN`), not `>= ADMIN`.
//! - `InstanceAdminPermission` matches `role__gte=15` with no
//!   active/verified filter on the row.
//! - Runner visibility has a single value (`PRIVATE=0`); anything else
//!   denies by default.
//! - `request_is_desktop` treats a broken session store as "not desktop".

use pidash_types::WorkspaceId;

use crate::scope::TenantScope;

pub mod allow;
pub mod desktop;
pub mod instance;
pub mod page;
pub mod project;
pub mod runner;
pub mod workspace;

/// Workspace/project role values (`ROLE_CHOICES`: Admin=20, Member=15,
/// Guest=5, in `pi_dash/db/models/workspace.py` and `project.py`).
pub const ROLE_ADMIN: i32 = 20;
/// Workspace/project role values (`ROLE_CHOICES`: Admin=20, Member=15,
/// Guest=5, in `pi_dash/db/models/workspace.py` and `project.py`).
pub const ROLE_MEMBER: i32 = 15;
/// Workspace/project role values (`ROLE_CHOICES`: Admin=20, Member=15,
/// Guest=5, in `pi_dash/db/models/workspace.py` and `project.py`).
pub const ROLE_GUEST: i32 = 5;

/// Roles of `ROLE` (`app/permissions/base.py`, `db/models/project.py`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    Admin,
    Member,
    Guest,
}

impl Role {
    /// Numeric value stored in the `role` column.
    pub fn value(self) -> i32 {
        match self {
            Role::Admin => ROLE_ADMIN,
            Role::Member => ROLE_MEMBER,
            Role::Guest => ROLE_GUEST,
        }
    }

    /// Map a stored value back; unknown values (possible only through raw
    /// SQL, since the column has choices) match no role list.
    pub fn from_value(value: i32) -> Option<Role> {
        match value {
            ROLE_ADMIN => Some(Role::Admin),
            ROLE_MEMBER => Some(Role::Member),
            ROLE_GUEST => Some(Role::Guest),
            _ => None,
        }
    }
}

/// DRF safe methods (`rest_framework.permissions.SAFE_METHODS`).
pub fn is_safe_method(method: &str) -> bool {
    matches!(method, "GET" | "HEAD" | "OPTIONS")
}

/// Deny unless `facts_workspace` is the scope's workspace.
///
/// Every Python rule filters its membership rows by workspace slug, so the
/// caller passes the slug its SQL filtered on; a mismatch means the facts
/// belong to another workspace and the check fails closed.
pub fn scope_allows(scope: &TenantScope, facts_workspace: &WorkspaceId) -> bool {
    scope.workspace_id() == facts_workspace
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn role_values_match_django_choices() {
        assert_eq!(Role::Admin.value(), 20);
        assert_eq!(Role::Member.value(), 15);
        assert_eq!(Role::Guest.value(), 5);
    }

    #[test]
    fn unknown_role_value_matches_nothing() {
        assert_eq!(Role::from_value(0), None);
        assert_eq!(Role::from_value(20), Some(Role::Admin));
    }

    #[test]
    fn safe_methods_match_drf() {
        for method in ["GET", "HEAD", "OPTIONS"] {
            assert!(is_safe_method(method), "{method}");
        }
        for method in ["POST", "PUT", "PATCH", "DELETE"] {
            assert!(!is_safe_method(method), "{method}");
        }
    }

    #[test]
    fn cross_workspace_facts_deny() {
        let scope = TenantScope::new(WorkspaceId::from("ws-a"));
        assert!(scope_allows(&scope, &WorkspaceId::from("ws-a")));
        assert!(!scope_allows(&scope, &WorkspaceId::from("ws-b")));
    }
}
